### ISSUE TITLE
Direction collapse on StockNet persists post-#6-fix (still constant-up)

### ISSUE THREAD

**Issue by dgonier** [dgonier/consciousness-ecology-codd#8]
## Status
PROPOSED

## Motivation
The #6 architectural fix (input-conditioned hunter_state) recovered ticker / magnitude / horizon conditioning, visible on dev (+11.9%) and held-out (+11.4%) rule-based reward. But the **direction** field on StockNet (#5) post-fix is *bit-identical* to pre-fix:

```
Pre-fix  (ipo_seed7):  ACCURACY=0.720, MCC=0.000, TP=36 TN=0 FP=14 FN=0
Post-fix (ipo_seed8):  ACCURACY=0.720, MCC=0.000, TP=36 TN=0 FP=14 FN=0
```

The model still predicts "up" on every one of 50 StockNet days. Same constant-class collapse. The #6 fix didn't touch this pathology — meaning **direction collapse is a separate failure mode**, not a downstream effect of the ticker collapse.

## Proposal

Three plausible root causes, ranked by how cheap they are to test:

### 1. Reward-weight bias in training (cheap to test)
- `reward_prediction()` weights: ticker=0.40, direction=0.25, magnitude=0.15, horizon=0.10, sigma=0.05, confidence=0.05
- Direction is binary — getting it right earns 0.25, getting it wrong earns 0.0. **There's no penalty for "abstain"-on-direction beyond losing 0.25.**
- During SFT, the model is teacher-forced to the target direction, so SFT shouldn't bias toward "up" — but during IPO sampling, the model may discover that emitting "up" is locally near-optimal because:
  - "up" matches a slight majority of training scenarios (~52% up given current synthetic distribution)
  - Rejected samples may include more "down" predictions that scored low for unrelated reasons (wrong ticker), so IPO learns "down → bad" indirectly
- **Test**: Compute the up/down distribution in the IPO sampling outputs vs targets. If sampled rejections lean "down" disproportionately, IPO is creating a spurious correlation.

### 2. Decode-time greediness (cheap to test)
- Decoding may be using greedy or low-temperature sampling, which amplifies the "up" prior at every token-emit step.
- **Test**: Re-run StockNet smoke with temperature=1.0 sampling instead of greedy. If MCC > 0, decoding bias is the issue.

### 3. Genuine architectural blind spot (expensive to fix)
- The predator's Channel attention may not have enough signal to differentiate "up-evidence" from "down-evidence" cleanly, even with the input-conditioned Q.
- **Test**: train a third seed (seed 9) with a simple direction-rebalancing loss term (e.g., entropy regularization on the direction logits) and see if MCC moves.

## Why this isn't "just rerun with a different prompt"
The collapse is robust across two seeds (7, 8), two reward landscapes (judge / rule-based), and two training configs (SFT-only vs SFT+IPO). It's structural. The fix path needs to address the root cause, not the surface symptom.

## Context pointers
- Files: `trophic/training/xml_schema.py:reward_prediction()` — reward weights
- Files: `trophic/training/ipo.py` — sampling logic
- Files: `scripts/eval_stocknet.py` — eval harness
- Logs: `logs/stocknet_smoke_ipo_seed7.log`, `logs/stocknet_smoke_ipo_seed8.log` — both show TP=36 TN=0 FP=14 FN=0
- Related issue: #6 (ticker collapse — fixed; same root cause investigation methodology applies here)

## Dependencies
- Blocked by: nothing (can start immediately)
- Blocks: paper-go decision (need MCC > 0 for the StockNet claim to be honest)
- Independent of: #3 (LangGraph), #2 (Director Nodes), #1 (slow nodes)

## Trigger conditions
Start when (a) post-fix architecture is otherwise stable (it is), AND (b) the rule-based dev/held-out gains have been confirmed under multi-seed (currently single-seed). Multi-seed first ensures we're not chasing a seed-specific pathology.

## Decision criteria
- StockNet MCC ≥ 0.10 on the same top-5 × 10 days subsample
- Per-ticker direction accuracy ≥ 0.55 across all 5 (currently AAPL=0.60 from constant-up; want all tickers ≥ 0.55 with non-trivial down predictions)
- Distribution of predictions on the eval set: at least 30% "down" decisions (currently 0%)

## Open questions
- Is there asymmetry in the training data that biases toward "up"? Quick check: count up/down in training set targets.
- Does the held-out test set's better-than-random performance (0.372 reward) come from ticker matches alone, or from any direction signal?

## Effort estimate
- Cheap tests (reward audit + decode temperature sweep): ~3-4 hours
- Multi-seed validation: ~6-8 hours per seed (one round-trip SFT + IPO)
- Direction-rebalancing training: ~6 hours
- Total: 1-2 days

## Out of scope
- Architectural overhaul of the predator (the architecture works on every other field; direction is a fixable symptom)
- New benchmarks (StockBench is queued in todo/02 but waits on this)

**Comment by dgonier** [comment]
Spawned from #6 after the architectural fix landed. The fix validated on rule-based dev/held-out (+11-12%) but did NOT move StockNet MCC off zero. Direction collapse is now confirmed as a distinct pathology.

**Comment by dgonier** [comment]
## Diagnostic update: 2 of 3 root causes ruled out, decode-greediness now lead candidate

### DIAG 1: Training-distribution direction prior — RULED OUT
- Training: 48.4% up / 45.6% down (balanced)
- Held-out: 52.9% up / 47.1% down (balanced)
- StockNet test slice: 72% up (this explains the 0.72 acc but is data-side, not model-side)

### DIAG 1d: Predictions on balanced held-out — confirms collapse is robust
- Held-out targets: 36 up, 32 down
- **Held-out predictions: 68 up, 0 down (100.0% constant-up)**
- Confusion: `up→up: 36, down→up: 32`
- The model emits "up" on every single one of 68 held-out scenarios despite the input being balanced. **Collapse is in the policy, not the data.**

### DIAG 3: IPO rejection-pool bias — RULED OUT
- 25 sampled rejected completions: 11 up / 11 down / 3 other
- Chosen targets: 51.5% up / 48.5% down
- IPO is *not* teaching "down → bad" by sampling artifact

### DIAG 2: Decode-temperature sweep — IN PROGRESS
- Found: `model_host.py:444 do_sample=False, temperature=N/A` — decode is hardcoded greedy.
- Added `TROPHIC_DECODE_TEMP` env var to enable sampling.
- Running sweep on ipo_seed8_best.pt at T=0.5, 1.0, 1.5 (each ~25 min). Log: `logs/stocknet_temp_sweep.log`.

### Lead hypothesis
Greedy decoding (`do_sample=False`) at every direction-token-emit step amplifies the policy's slight up-prior, locking in constant-up. With sampling enabled, the model has a chance to surface the down signal that's encoded in the policy distribution but suppressed by argmax. **If MCC > 0 at any temperature, decode greediness is the cause and the fix is trivial: enable sampling for direction-bearing fields.**

If MCC stays at 0 across all temperatures → the policy distribution itself is constant-up and we go to Tier 2-3 fixes (reward-shaping, entropy regularization, or Bayesian Predator).

**Comment by dgonier** [comment]
Related: #9 — proposed EnvironmentStream rebuild as the architecturally-correct path for real-world data integration. If temperature sweep confirms Outcome B (policy distribution itself is constant-up), #9 becomes the fix path; the StockNet integration is the forcing function for getting the architecture right rather than a quick patch.

**Comment by dgonier** [comment]
Related: #10 — if direction collapse persists post-#9 (EnvironmentStream rebuild), the next fix path is per-species CPT. Synthetic balanced scenarios + uniform Qwen3-4B base may not have enough signal differentiation for the predator to learn direction; CPT'd species pretrained on real disjoint finance corpora would change the input distribution at the deepest level.

**Comment by dgonier** [comment]
ENVSTREAM rebuild (#9) landed; partial training run did **not** move MCC off zero on this checkpoint.

**Status:** unchanged. Direction collapse persists in the same shape as before — predator emits constant `<prediction><ticker>GOOG</ticker></prediction>` on every dev scenario (rule-based reward = 0.050, parse_floor only).

**Caveat — this is NOT a fair test of the hypothesis.** SFT seed 9 was truncated at step 100 (best eval_loss=1.142) because the in-loop `eval_decode` over 14 scenarios was running 7-15 minutes per eval round and blew the time budget. Compare to seed 8 which trained to step 600 with eval_loss=0.192. IPO seed 9 was killed before its first eval at step 100 with no IPO checkpoint saved. So the eval cascade ran against a partial SFT checkpoint, not a properly-trained EnvStream-mode checkpoint.

**Follow-up plan:**
1. Profile / fix the in-loop eval-decode bottleneck — `model.generate(do_sample=False)` is hitting `max_new_tokens` cap when the model is producing un-trained gibberish. Likely fix: drop in-loop decoded eval from SFT, replace with eval_loss only; do decode-eval at end of SFT or via `eval_xml_checkpoint.py` script post-hoc. Alternatively reduce `max_new_tokens` to ~96 during in-loop eval.
2. Re-run `sft_seed9` (full 600 steps) and `ipo_seed9` (full 800 steps) with `TROPHIC_USE_ENVSTREAM=1` enabled.
3. Then run the eval cascade: dev + held-out + StockNet smoke. Capture MCC as the decision metric.

Decision criterion stays the same: MCC > 0.05 = success; > 0.10 = paper-grade. We didn't get to test it on this run.

Other diagnostic info from the truncated run:
- Mock SFT smoke (`TROPHIC_MOCK_MODELS=1 TROPHIC_MOCK_PREDATOR=1 TROPHIC_STEPS=10`): exit 0
- EnvStream cortical-column fan-out diag: validates wavelength filter mechanics (TickDelta/Anomaly attend ohlcv; SocialSignal attends tweets)
- 83/83 tests passing, no regression introduced by the architectural rewrite

Issue stays open.

**Comment by dgonier** [comment]
Dataset research suggests #10 phase 1 should target Interrogator (FinCoT) FIRST, not Disclosure. The diagnosis: direction collapse is a *reasoning* failure dressed as a *prediction* failure. FinCoT is the only candidate dataset that installs explicit, verifiable, multi-step conditioning. If FinCoT-trained Interrogator gating Predator output breaks #8, that's the paper claim. If not, the failure is deeper than reasoning. See #10 comment for the full re-prioritization.

**Comment by dgonier** [comment]
Related: #11 — FinCoT prompting (arXiv 2506.16123) is a Tier-0 prompt-only probe. If it moves StockNet MCC on existing ipo_seed8_best.pt without retraining, structural reasoning is enough to break the collapse and #10 becomes paper-strengthening rather than paper-blocking.

**Comment by dgonier** [comment]
## EnvironmentStream architecture (#9) does NOT solve #8

ipo_seed9_best.pt (post-EnvStream) StockNet smoke: ACCURACY=0.720, MCC=0.000, TP=36 TN=0 FP=14 FN=0 — **bit-identical** to ipo_seed7 and ipo_seed8. Predator still emits constant "up" on every one of 50 StockNet days.

The architecture lifts dev (0.417, all-time best) and judge eval (0.607, all-time best, reached 4× faster) but the policy's output distribution at decode time hasn't changed.

Verdict: input-substrate architecture is orthogonal to direction collapse. The bottleneck is **reasoning structure** (force the model to condition on input via CoT) or **pretraining distribution** (give the model real direction-prediction priors), not the input pipeline.

Routes:
- #11 (FinCoT prompting) — running NOW on ipo_seed8_best.pt as Experiment A. Pure prompt-only intervention. Result in ~25 min.
- #10 (per-species CPT) — Interrogator-on-FinCoT-dataset first if #11 doesn't move the needle.

**Comment by dgonier** [comment]
## Three hypotheses tested for #8; all three failed to move MCC

After EnvironmentStream (#9) and FinCoT prompting (#11) both failed to break the constant-up direction collapse on StockNet, the remaining intervention is **#10 per-species continued pretraining** with the dataset deep-dive's recommendation: Interrogator-on-FinCoT-dataset first (highest-leverage per the research).

### Summary of attempts

| Issue | Hypothesis | Verdict | StockNet MCC |
|---|---|---|---|
| #6 | Predator Q was input-independent (`role_prefix.mean()` alone) | **Confirmed + fixed**; helped dev/held-out but not direction | 0.000 |
| #9 | Input substrate (EnvStream) gives producers richer K/V to discriminate | **Falsified for #8**; lifted training metrics but MCC unchanged | 0.000 |
| #11 | Structured-CoT prompting (FinCoT paper, arXiv 2506.16123) gates the policy through reasoning steps | **Untestable cheaply**; longer role_prefix breaks trained Channels (#12) | undefined (gibberish) |

### What's left
- **#10 per-species CPT** — Interrogator on FinCoT dataset (Fino1 reasoning paths). This bakes structured reasoning into the training distribution, sidestepping #12's prompt-length fragility, and is the highest-leverage intervention per the dataset deep-dive.
- **#12 Q-robustness fix** — could come first (so #10's training is robust) OR after (#10 cooks the fix in by training with the new structured prompt format).

### Recommended order
1. Open #12 (prompt-length-fragility fix) — small, ~3 hr
2. Apply Option 2 (normalize role_q to fixed norm) and retrain a sanity-check seed to confirm robustness
3. Then start #10 phase 1 (Interrogator-on-FinCoT) on the now-robust architecture
4. Re-test FinCoT prompting (#11 reopened) on the robust architecture as a cheap verification before committing to multi-species CPT

This sequencing means the next chunk of work is ~1 day of code + 1 retrain (to land #12), then 1-2 days for #10 phase 1.

**Comment by dgonier** [comment]
## All cheap interventions exhausted; #8 requires co-training

Five interventions tested for #8:
1. **#6 predator-Q fix** — helped dev/held-out (+12%) but not direction
2. **#9 EnvironmentStream** — best dev/judge ever (0.417 / 0.607) but MCC=0.000
3. **#11 FinCoT prompting (no retrain)** — gibberish (#12 fragility)
4. **#12 Q-normalization** — defensive fix; doesn't address the deeper Channel↔base fragility
5. **#10 phase 1 partial LoRA + ipo_seed8** — schema collapse (well-formed XML, no `<direction>`)

The pattern: the architecture has Channels trained against vanilla Qwen3-4B hidden states. ANY modification to the base model (LoRA, prompt change, etc.) shifts those hidden states out-of-distribution and the trained Channels produce something well-formed but wrong. There's no shortcut — fixing the base requires retraining the Channels with the new base in place.

The honest path forward: ONE multi-hour run with all the architectural pieces co-trained:
- FinCoT LoRA active throughout (~3 hr LoRA training)
- SFT seed 12 with LoRA active (~85 min)
- IPO seed 12 with LoRA active (~75 min)
- Eval cascade (~50 min)

Total: ~6 hr GPU, single seed. If MCC moves with this, the architecture works as designed and the fix is reproducible. If not, the bottleneck is genuinely deeper than reasoning installation and we're at the multi-species CPT level.

For now, leaving #8 open. The work to actually test this is real but bounded.
